### PR TITLE
USWDS - Utilities: Custom breakpoints

### LINKED ISSUE #5984: USWDS - Feature: Add theme setting for custom breakpoints
### Is your feature request related to a problem? Please describe.

When the USWDS system is integrated into an existing site, there are sometimes additional breakpoints already integrated that do not exist in the USWDS system. This means that the USWDS utilities are incomplete because the system doesn't allow for adding additional breakpoints.

### Describe the solution you'd like

I would like to be able to use the USWDS system breakpoints as well as optionally add a custom breakpoint for generating utility classes.

### Describe alternatives you've considered

_No response_

### Additional context

The VA.gov system uses a breakpoint called `medium-screen` which is 768px (or 48rem). This breakpoint does not exist in the USWDS system but we would like to continue to use it along with the other breakpoints USWDS offers. This breakpoint is being used in [167+ files in VA.gov](https://github.com/search?q=repo%3Adepartment-of-veterans-affairs%2Fvets-website%20medium-screen&type=code).

### Code of Conduct

- [X] I agree to follow this project's [Code of Conduct](https://designsystem.digital.gov/about/community/#community-conduct).
- [X] I checked the [current issues](https://github.com/uswds/uswds/issues?q=is%3Aissue+is%3Aopen+label%3A%22Type%3A+Feature+Request%22) for duplicate feature requests.

### PR DESCRIPTION
**Pull request by mahoneycm**
# Summary

*Continuation of #5985. Thank you for your contribution @jamigibbs.*

**Added new `$theme-utility-breakpoints-custom` setting.** This setting generates responsive variants of USWDS utility classes at custom breakpoints.

Being able to add a custom breakpoint to the utilities is useful for integrating USWDS with existing frameworks or other design requirements. Custom breakpoints will generate utilities alongside the default system breakpoints.

## Sample implementation
To generate a custom breakpoint for your USWDS utilities, add custom breakpoint values to a [Sass map](https://sass-lang.com/documentation/values/maps/) for `$theme-utility-breakpoints-custom` in your [USWDS settings configuration](https://designsystem.digital.gov/documentation/settings/#configuring-custom-uswds-settings). Here is an example implementation:

```scss
@use "uswds-core" with (
  $theme-utility-breakpoints-custom: (
    "new-breakpoint": 800px
  )
);
```

> [!Important]
> Custom breakpoints must be defined with `px` values. 

> [!Tip]
> Every breakpoint added to the system adds about 50kb to the compiled CSS. Be careful to only turn on the breakpoints you need for your project. You can turn off default USWDS breakpoints with the `$theme-utility-breakpoints` setting. More details about `$theme-utility-breakpoints` can be found  on the [Settings page](https://designsystem.digital.gov/documentation/settings/#utiltiies-settings). 

## Breaking change

This is not a breaking change.

## Related issue

Closes #5984

## Related pull requests

Continuation of #5985

[Site documentation update →](https://github.com/uswds/uswds-site/pull/2837)

## Preview link

[Preview link →](https://federalist-3b6ba08e-0df4-44c9-ac73-6fc193b0e19c.sites.pages.cloud.gov/preview/uswds/uswds/cm-custom-breakpoints/)

## Problem statement

It is currently not possible for users to implement custom breakpoints and use them alongside our utility classes

## Solution

Create new scss map for users to define their own breakpoints and build out utility classes with.

## Major changes

- Created `$theme-utility-breakpoints-custom` map for users to name and define their own breakpoints
- Created `px-to-user-em()` sass function
- Collect the breakpoints in system properties
- Update `_utility-builder.scss` to take custom breakpoints
- Update `at-media` mixin to create utility classes with custom breakpoints
- At this time, breakpoints can only be set in `px` units.
- Custom breakpoints follow the same `em` conversation methods as standard USWDS breakpoints
  - If `$theme-respect-user-font-size` is `true`, breakpoints are converted to `em`
  - If `$theme-respect-user-font-size` is `false`, breakpoints are output as `px`

### Testing steps

1. Checkout feature branch
2. Visit `settings-utilities.scss` and add custom breakpoints with `px` and `rem`  (or some other unit)
    - Ex:
    
    ```scss
    $theme-utility-breakpoints-custom (
      "bp-px": 64px;
      "bp-rem": 10rem;
    )
    ```
    
3. Run `npx gulp compileSass`
4. Confirm the non `px` breakpoint triggers Warning message in the terminal.
5. View output USWDS css at `dist/css/uswds.css`.
6. Search for your custom breakpoint by name.
7. Confirm new media queries are built using values of your custom breakpoints.
8. Confirm value is correctly converted to `em`.
9. Confirm new utility classes are built using the name of your breakpoints.
10. Confirm non `px` breakpoints were not built.
11. Open `_settings-typography` and set  `$theme-respect-user-font-size` to `false !default`
12. Recompile sass `npx gulp compileSass`
13. View output USWDS css at `dist/css/uswds.css` 
14. Confirm breakpoints are now output as `px`.

### Update responsive settings test

Turn on a responsive utility class that is disabled by default and confirm that the custom breakpoints are created for the new utility.

1. Visit `_settings-utilities.scss`
2. Update `background-color-settings-complete` change `responsive: false` to `responsive: true`
3. Compile sass `npx gulp compileSass`
4. View output USWDS css at `dist/css/uswds.css` 
5. Find your custom breakpoint and confirm `bg-(color)` utility classes are created for your custom breakpoint

### Testing checklist

- [ ] Approve map merge method
- [ ] Approve collecting custom breakpoints in `extended` system property
- [ ] Confirm approach in the new `px-to-user-em` sass function is valid.
- [ ] Confirm at-mixin and utility builder properly output breakpoints and utility classes
- [ ] Confirm warning message appears when non `px` value is used.
- [ ] Updated responsive settings respect new breakpoints.
- [ ]  No regressions updating standard breakpoints.

### CODE REVIEW

**Comment by mahoneycm** [comment]
## Additional considerations

This work currently **does not** build Layout Grid utility classes for the new custom breakpoints, but it would be possible to update `_usa-layout-grid.scss` to include the custom breakpoints.

Is that something we want to provide to the users with this update? Alternatively, it could be an additional enhancement after this work is merged.

**Comment by mahoneycm** [comment]
@jamigibbs After some discussion, we've opted to only take custom breakpoints in `px` since it feels like a standard for breakpoints and a common language between designers and developers. We wanted to get your team's opinion on this approach! Do you think this is a valid decision? Are there substantial reasons we should also take in `rem` or any other units?

I'm unable to tag you as a reviewer for this work, but if you could test this PR to see if it works with your custom breakpoints the way you'd like, it'd be greatly appreciated! All feedback is welcome 👍

File: packages/uswds-core/src/styles/functions/units/px-to-user-em.scss
Comment (mahoneycm) [inline packages/uswds-core/src/styles/functions/units/px-to-user-em.scss:21]: I opted not to re-invent the wheel and use functions we already had to convert pixels to rem.

Alternatively, we could use the following sass math conversion if we don't want to rely on the other functions:

```scss
  @if unit($pixels) != "px" {
    @error 'This value must be in pixels';
  }
  $px-to-user-em: (math.div($pixels, root.$root-font-size-equiv)) * 1em;
  $px-to-user-em: math.div(math.round($px-to-user-em * 100), 100);
```

File: packages/uswds-core/src/styles/functions/units/rem-to-user-em.scss
Comment (mahoneycm) [inline packages/uswds-core/src/styles/functions/units/rem-to-user-em.scss:16]: **note:** Added error to match other conversion functions

File: packages/uswds-core/src/styles/mixins/helpers/at-media.scss
Comment (mahoneycm) [inline packages/uswds-core/src/styles/mixins/helpers/at-media.scss:10]: **note:** Moved variable declaration out of mixin since they are used in both `at-media` and `at-media-max`

**Review by mahoneycm — COMMENTED** [review]
A couple notes on my approach here

**Comment by jamigibbs** [comment]
> @jamigibbs After some discussion, we've opted to only take custom breakpoints in `px` since it feels like a standard for breakpoints and a common language between designers and developers. We wanted to get your team's opinion on this approach! Do you think this is a valid decision? Are there substantial reasons we should also take in `rem` or any other units?
> 
> I'm unable to tag you as a reviewer for this work, but if you could test this PR to see if it works with your custom breakpoints the way you'd like, it'd be greatly appreciated! All feedback is welcome 👍

@mahoneycm Thanks Charlie! `px` will be fine for us. I originally called that out just for visibility but I don't see any reason why it needs to be anything other than `px`. 

I also tested your branch locally with our implementation and it seems to be working great! The custom breakpoints were generating and an error message displayed if I tried to using something other than `px`.

Screenshot 2024-09-18 at 2 55 37 PM
 
Screenshot 2024-09-18 at 2 18 23 PM

**Review by mejiaj — APPROVED** [review]
Thanks @mahoneycm. I've been able to successfully test following your instructions. I've created a branch on uswds-sandbox `test-uswds-breakpoints-6048`, so others can test as well. Links to that testing environment:
- **Branch:** [`test-uswds-breakpoints-6048`](https://github.com/uswds/uswds-sandbox/tree/test-uswds-breakpoints-6048).
- **Draft PR:** uswds/uswds-sandbox#153
- **Preview:** [Sandbox prose →](https://federalist-d5e7c07c-6ffa-4b8e-935f-49a7df24a505.sites.pages.cloud.gov/preview/uswds/uswds-sandbox/test-uswds-breakpoints-6048/pages/prose/)

**Confirming** 
- [x] New utils are generated for custom breakpoint
- [x] Breakpoint is available via `at-media`
- [x] Warning on passing non-px values
- [x] Additional utils are generated
- [x] Overriding root font size sets media queries with `px`

### Non-pixel values

Confirming that passing non-px values give a warning.

**Input**
```scss
// _uswds-theme.scss
@use "uswds-core" with (
  $theme-utility-breakpoints-custom: (
    "desktop-custom-rem": 70rem // 1120px,,
  )
);
```

**Output**
```sh
Warning: Custom breakpoint `desktop-custom-rem` has the unit rem. Custom breakpoints must be set in px.
    ../../../@uswds/uswds/packages/uswds-core/src/styles/mixins/helpers/at-media.scss 35:7  at-media()
    ../../../../src/_styles/_uswds-theme-custom-styles.scss 11:1                            @forward
    ../../../../src/_styles/styles.scss 3:1   
```

### Additional utils

> [!WARNING]
> Adding responsive variants has an impact on CSS compile size.

Confirming that additional utilities are output.

**Input**
```scss
// _uswds-theme.scss
@use "uswds-core" with (
  $theme-utility-breakpoints-custom: (
    "desktop-custom": 1100px,
  ),
  $background-color-settings: (
    responsive: true,
  )
);
```

**Output**

```css
/* styles.css */
@media all and (min-width: 68.75em) {
  .desktop-custom\:bg-transparent,
  .desktop-custom\:hover\:bg-transparent:hover {
    background-color: transparent;
  }
  .desktop-custom\:bg-black,
  .desktop-custom\:hover\:bg-black:hover {
    background-color: #000;
  }
  /* … */
}
```

### Overriding root font size

**Input**
```scss
// _uswds-theme.scss
@use "uswds-core" with (
  $theme-utility-breakpoints-custom: (
    "desktop-custom": 1100px,
  ),
  $theme-respect-user-font-size: false,
  $theme-root-font-size: 20px,
);
```

**Output**
```css
/* styles.css */
@media all and (min-width: 1100px) {
  /* … */
}
```

### Output CSS size

Output CSS file size after adding a single custom breakpoint. This issue is compounded when we're adding responsive variants on top.

| File       | Before | After |
| :--------- | :----- | ----- |
| styles.css | 524KB  | 572KB |

We should create an issue to support `at-media` **without** generating utils for performance.

**Review by amyleadem — COMMENTED** [review]
Looking good @mahoneycm! (And thanks for your original work on this @jamigibbs!) 

> [!important]
> Before I approve, I wanted to highlight that this PR does _not_ generate custom responsive breakpoints for layout grid or icon-list-size classes. I have added additional details in the notes below.
> 
> Are we comfortable with this moving forward without these responsive classes?

Everything else is looking good!

## Testing details

I tested in uswds-sandbox using the following config:
```scss
@use "uswds-core" with (
  $theme-show-compile-warnings: false,
  $theme-show-notifications: false,
  $theme-utility-breakpoints-custom: (
    "new-breakpoint": 800px
  ),
  $theme-utility-breakpoints: (
    widescreen: true,
  ),
  $left-settings: (
    responsive: true
  )
);
```

- [ ] Confirm the `$theme-utility-breakpoints-custom`  setting generates new style rules for each new pixel-based value in the map
    - Note: There are some gaps in what `$theme-utility-breakpoints-custom` generates vs `$theme-utility-breakpoints`
        - Turning on `new-breakpoint` in ` $theme-utility-breakpoints-custom` generates 898 new references to `.new-breakpoint`in the CSS. Turning on `widescreen` in `$theme-utility-breakpoints` generates 1060 new references to `.widescreen\:`
        - After some digging, discovered that difference was that the custom breakpoint was not generated for:
            - usa-icon-list sizes - for example, it did not generate `.new-breakpoint\:usa-icon-list—size-1`, but did generate `.widescreen\:usa-icon-list—size-1` (In total, there were 90 references for `usa-icon-list--size` generated for `widescreen`)
            - layout grid classes - For example, it did not generate `.new-breakpoint\:grid-container` but did generate `.widescreen\:grid-container`  (In total, 72 grid references were generated for `widescreen`)
        - For reference, here is the difference in CSS size:
            - with default breakpoints: 524kb
            - with only a new custom breakpoint: 572kb
            - with only an additional standard breakpoint (`widescreen`): 587kb
- [x] Confirm there is an appropriate warning in the console when a non-px unit is added to the `$theme-utility-breakpoints-custom` setting map
    - Note: I added a non-breaking suggestion below to add a bit more clarity to the error message
- [x] Confirm that setting `$theme-respect-user-font-size: false` outputs new breakpoint in `px`
- [x] Confirm that setting `$theme-respect-user-font-size: true` outputs new breakpoint in `em`
- [x] Confirm conversion from px to em is accurate (eg, 800px outputs as 50em)
- [x] Confirm that turning on a responsive variant of a USWDS utility generates the custom breakpoint from `$theme-utility-breakpoints-custom` as well as the default breakpoints from `$theme-utility-breakpoints` (I tested `$left-settings`)
- [x] Confirm the code meets standards

File: packages/uswds-core/src/styles/mixins/helpers/at-media.scss
Comment (mejiaj) [reply packages/uswds-core/src/styles/mixins/helpers/at-media.scss]: Good point. We should update anyways because the test instructions in description say non-px values will trigger a warning. So we can remove the `either rem or px` note and just say `px`.

Comment (amyleadem) [reply packages/uswds-core/src/styles/mixins/helpers/at-media.scss]: Whoops, I realize my comment was made a lil while ago and is a bit outdated. It looks like it doesn't reference `rem` in the warning message anymore. I'll go ahead and make the update about the breakpoint not generating.

Comment (amyleadem) [reply packages/uswds-core/src/styles/mixins/helpers/at-media.scss]: Added in [f319030](https://github.com/uswds/uswds/pull/6048/commits/f31903066e5a572c4978d2102c5b138ac3e02fc4)

**Comment by amyleadem** [comment]
Update: I have opened PR  #6083 as a candidate to add support for layout grid utility classes to this PR.

**Comment by amyleadem** [comment]
Update: I have merged in #6083 after team approval. Will work on finalizing my review of this PR now.

File: packages/uswds-core/src/styles/mixins/helpers/at-media.scss
Comment (amyleadem) [inline packages/uswds-core/src/styles/mixins/helpers/at-media.scss:35]: **suggestion (non-blocking)**: Thinking it would be good to reference the related setting in the warning message. Small tweak, but I think it will make troubleshooting just a bit easier. 

Current error message:
image

```suggestion
      @warn 'Custom breakpoint `#{$quoted-bp}` was defined with the unit #{unit(map.get($custom-breakpoints, $quoted-bp))} and can not be generated. Values for `$theme-utility-breakpoints-custom` must be set with `px` units.';
```

File: packages/usa-layout-grid/src/styles/_usa-layout-grid.scss
Comment (amyleadem) [inline packages/usa-layout-grid/src/styles/_usa-layout-grid.scss:10]: > [!note]
> Flagging that layout grid and icon list size use [$system-breakpoints in breakpoints.scss](https://github.com/uswds/uswds/blob/develop/packages/uswds-core/src/styles/tokens/units/breakpoints.scss) and the other utilities build from [`$system-properties.breakpoints` in _properties.scss](https://github.com/uswds/uswds/blob/develop/packages/uswds-core/src/styles/_properties.scss#L157). Both seem to collect the same values:
>
> ```scss
> /*breakpoints.scss*/
> $system-breakpoints: general.map-collect(
>   map.get(spacing.$system-spacing, "large"),
>   map.get(spacing.$system-spacing, "larger"),
>   map.get(spacing.$system-spacing, "largest")
> );
> ```
> 
> ```scss
> /*_properties.scss*/
>   breakpoints: (
>     standard:
>       map-collect(
>         map.get($system-spacing, "large"),
>         map.get($system-spacing, "larger"),
>         map.get($system-spacing, "largest")
>       ),
>     extended: (),
>   ),
> ```
> 
> To keep things scoped, we have decided to use `$system-breakpoints` here for now, but can explore replacing it in a future follow-on issue. 
> 
> More details about this can be found in this [comment thread](https://github.com/uswds/uswds/pull/6083/files#r1777767472).

**Review by amyleadem — APPROVED** [review]
Looks good to me! I do have one suggestion to add detail to the warning message in the comment below, but I don't consider it to be a blocker

**Review by thisisdano — APPROVED** [review]
Nice feature. Thank you @jamigibbs and @mahoneycm!

**Pull request merged**
